File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/BadgeLayoutSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import eu.pretix.libpretixsync.api.ApiException;
import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.db.BadgeLayout;
import eu.pretix.libpretixsync.db.Item;
import eu.pretix.libpretixsync.utils.HashUtils;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;
import io.requery.query.Tuple;
import io.requery.util.CloseableIterator;

public class BadgeLayoutSyncAdapter extends BaseDownloadSyncAdapter<BadgeLayout, Long> {

    public BadgeLayoutSyncAdapter(BlockingEntityStore<Persistable> store, FileStorage fileStorage, String eventSlug, PretixApi api, String syncCycleId, SyncManager.ProgressFeedback feedback) {
        super(store, fileStorage, eventSlug, api, syncCycleId, feedback);
    }

    @Override
    public void updateObject(BadgeLayout obj, JSONObject jsonobj) throws JSONException {
        obj.setEvent_slug(eventSlug);
        obj.setIs_default(jsonobj.getBoolean("default"));
        obj.setServer_id(jsonobj.getLong("id"));
        obj.setJson_data(jsonobj.toString());

        String remote_filename = jsonobj.optString("background");
        if (remote_filename != null && remote_filename.startsWith("http")) {
            String hash = HashUtils.toSHA1(remote_filename.getBytes());
            String local_filename = "badgelayout_" + obj.getServer_id() + "_" + hash + ".pdf";
            if (obj.getBackground_filename() != null && !obj.getBackground_filename().equals(local_filename)) {
                fileStorage.delete(obj.getBackground_filename());
                obj.setBackground_filename(null);
            }
            if (!fileStorage.contains(local_filename)) {
                try {
                    PretixApi.ApiResponse file = api.downloadFile(remote_filename);
                    OutputStream os = fileStorage.writeStream(local_filename);
                    InputStream is = file.getResponse().body().byteStream();
                    byte[] buffer = new byte[1444];
                    int byteread;
                    while ((byteread = is.read(buffer)) != -1) {
                        os.write(buffer, 0, byteread);
                    }
                    is.close();
                    os.close();
                    obj.setBackground_filename(local_filename);
                } catch (ApiException e) {
                    // TODO: What to do?
                    e.printStackTrace();
                } catch (IOException e) {
                    // TODO: What to do?
                    e.printStackTrace();
                    fileStorage.delete(local_filename);
                }
            } else {
                obj.setBackground_filename(local_filename);
            }
        } else {
            if (obj.getBackground_filename() != null) {
                fileStorage.delete(obj.getBackground_filename());
                obj.setBackground_filename(null);
            }
        }
    }

    @Override
    protected void prepareDelete(BadgeLayout obj) {
        super.prepareDelete(obj);
        if (obj.getBackground_filename() != null) {
            fileStorage.delete(obj.getBackground_filename());
        }
    }

    @Override
    public CloseableIterator<BadgeLayout> runBatch(List<Long> ids) {
        return store.select(BadgeLayout.class)
                .where(BadgeLayout.EVENT_SLUG.eq(eventSlug))
                .and(BadgeLayout.SERVER_ID.in(ids))
                .get().iterator();
    }

    @Override
    CloseableIterator<Tuple> getKnownIDsIterator() {
        return store.select(BadgeLayout.SERVER_ID)
                .where(BadgeLayout.EVENT_SLUG.eq(eventSlug))
                .get().iterator();
    }

    @Override
    String getResourceName() {
        return "badgelayouts";
    }

    @Override
    Long getId(JSONObject obj) throws JSONException {
        return obj.getLong("id");
    }

    @Override
    Long getId(BadgeLayout obj) {
        return obj.getServer_id();
    }

    @Override
    BadgeLayout newEmptyObject() {
        return new BadgeLayout();
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/BadgeLayoutSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import app.cash.sqldelight.db.QueryResult
import eu.pretix.libpretixsync.api.ApiException
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.sqldelight.BadgeLayout
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import eu.pretix.libpretixsync.utils.HashUtils
import org.json.JSONObject
import java.io.IOException

class BadgeLayoutSyncAdapter(
    db: SyncDatabase,
    fileStorage: FileStorage,
    eventSlug: String,
    api: PretixApi,
    syncCycleId: String,
    feedback: ProgressFeedback?,
) : BaseDownloadSyncAdapter<BadgeLayout, Long>(
    db = db,
    api = api,
    syncCycleId = syncCycleId,
    eventSlug = eventSlug,
    fileStorage = fileStorage,
    feedback = feedback,
) {
    override fun getResourceName(): String = "badgelayouts"

    override fun getId(obj: BadgeLayout): Long = obj.server_id!!

    override fun getId(obj: JSONObject): Long = obj.getLong("id")

    override fun getJSON(obj: BadgeLayout): JSONObject = JSONObject(obj.json_data!!)

    override fun queryKnownIDs(): MutableSet<Long>? {
        val res = mutableSetOf<Long>()
        db.badgeLayoutQueries.selectServerIdsByEventSlug(event_slug = eventSlug).execute { cursor ->
            while (cursor.next().value) {
                val id =
                    cursor.getLong(0) ?: throw RuntimeException("server_id column not available")
                res.add(id)
            }

            QueryResult.Unit
        }

        return res
    }

    override fun insert(jsonobj: JSONObject) {
        val backgroundFilename = processBackground(jsonobj, null)

        db.badgeLayoutQueries.insert(
            background_filename = backgroundFilename,
            event_slug = eventSlug,
            is_default = jsonobj.getBoolean("default"),
            json_data = jsonobj.toString(),
            server_id = jsonobj.getLong("id"),
        )
    }

    override fun update(obj: BadgeLayout, jsonobj: JSONObject) {
        val backgroundFilename = processBackground(jsonobj, obj.background_filename)

        db.badgeLayoutQueries.updateFromJson(
            background_filename = backgroundFilename,
            event_slug = eventSlug,
            is_default = jsonobj.getBoolean("default"),
            json_data = jsonobj.toString(),
            id = obj.id,
        )
    }

    private fun processBackground(jsonobj: JSONObject, oldFilename: String?): String? {
        val remote_filename = jsonobj.optString("background")
        var result: String? = null

        if (remote_filename != null && remote_filename.startsWith("http")) {
            val hash = HashUtils.toSHA1(remote_filename.toByteArray())
            val local_filename = "badgelayout_" + jsonobj.getLong("id") + "_" + hash + ".pdf"
            if (oldFilename != null && oldFilename != local_filename) {
                fileStorage.delete(oldFilename)
                result = null
            }
            if (!fileStorage.contains(local_filename)) {
                try {
                    val file = api.downloadFile(remote_filename)
                    val os = fileStorage.writeStream(local_filename)
                    val `is` = file.response.body!!.byteStream()
                    val buffer = ByteArray(1444)
                    var byteread: Int
                    while ((`is`.read(buffer).also { byteread = it }) != -1) {
                        os.write(buffer, 0, byteread)
                    }
                    `is`.close()
                    os.close()
                    result = local_filename
                } catch (e: ApiException) {
                    // TODO: What to do?
                    e.printStackTrace()
                } catch (e: IOException) {
                    // TODO: What to do?
                    e.printStackTrace()
                    fileStorage.delete(local_filename)
                }
            } else {
                result = local_filename
            }
        } else {
            if (oldFilename != null) {
                fileStorage.delete(oldFilename)
                result = null
            }
        }

        return result
    }

    override fun delete(key: Long) {
        db.badgeLayoutQueries.deleteByServerId(key)
    }

    override fun prepareDelete(obj: BadgeLayout) {
        super.prepareDelete(obj)
        if (obj.background_filename != null) {
            fileStorage.delete(obj.background_filename)
        }
    }

    override fun runInTransaction(body: TransactionWithoutReturn.() -> Unit) {
        db.badgeLayoutQueries.transaction(false, body)
    }

    override fun runBatch(parameterBatch: List<Long>): List<BadgeLayout> =
        db.badgeLayoutQueries.selectByServerIdListAndEventSlug(
            server_id = parameterBatch,
            event_slug = eventSlug,
        ).executeAsList()

}
